### PR TITLE
Architecture: Cleanup accessibility flags

### PR DESCRIPTION
**Pull request by talekhinezh**
## Summary
Use application layer flags to specify if a vault is freezable or not and whether a validator is stakeable or not.

## Details
* Remove `attach_access_rules`
* Add `stake_as_owner` validator method
* Add `BlueprintType` specifying properties whether an object is an inner or outer object
* Remove two levels of auth
* Add `MethodPermission::OuterObjectOnly`
* Add `VaultFrozenFlag`
* Add `vault_freeze`, `vault_recall`, `mint`, and `burn` resource manager features

### For dApp Developers
Added `stake_as_owner` as the method a validator owner may always be able to call in order to stake.

### For Internal Integrators
Validator and Vault flag substates have been added

### CODE REVIEW

File: radix-engine-interface/src/blueprints/resource/access_rules.rs [hunk 1]
```diff
@@ -69,6 +56,7 @@ impl MethodEntry {
 #[derive(Debug, Clone, PartialEq, Eq, Hash, Ord, PartialOrd, ScryptoSbor, ManifestSbor)]
 pub enum MethodPermission {
     Public,
+    OuterObjectOnly,
```
Comment (iamyulong) [inline radix-engine-interface/src/blueprints/resource/access_rules.rs:59]: As discussed, we're going to rename `MethodPermission` to `MethodAccessibility`, understanding there is overhead to downstream clients. Let's also add some comments.

File: radix-engine-interface/src/types/blueprint.rs [hunk 2]
```diff
@@ -10,6 +10,20 @@ use sbor::rust::prelude::*;
 use scrypto_schema::{InstanceSchema, KeyValueStoreSchema};
 use utils::ContextualDisplay;
 
+#[derive(Debug, Clone, PartialEq, Eq, ScryptoSbor)]
+pub enum ObjectBlueprintInfo {
+    Inner { outer_object: GlobalAddress },
+    Outer { features: BTreeSet<String> },
+}
```
Comment (iamyulong) [inline radix-engine-interface/src/types/blueprint.rs:17]: Related to the other comments, maybe?
```rust
pub struct ObjectBlueprintInfo {
    blueprint_type: BlueprintType,
    features: BTreeSet<String>
}
```

File: radix-engine/src/blueprints/package/package.rs [hunk 3]
```diff
@@ -1015,6 +1035,19 @@ impl PackageNativePackage {
         // Build node init
         {
             for (blueprint, definition_init) in setup.blueprints {
+                match (
+                    &definition_init.blueprint_type,
+                    &definition_init.auth_config.method_auth,
+                ) {
+                    (_, MethodAuthTemplate::Static(..)) => {}
+                    (_, MethodAuthTemplate::NoAuth) => {}
```
Comment (iamyulong) [inline radix-engine/src/blueprints/package/package.rs:1043]: Suggest renaming `NoAuth` to `AllPublic`

File: radix-engine/src/blueprints/resource/fungible/fungible_vault.rs [hunk 4]
```diff
@@ -261,6 +259,55 @@ impl FungibleVaultBlueprint {
 
         Ok(())
     }
+
+    pub fn assert_not_frozen<Y>(api: &mut Y) -> Result<(), RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        if !api.actor_is_feature_enabled(VAULT_FREEZE_FEATURE)? {
```
Comment (iamyulong) [inline radix-engine/src/blueprints/resource/fungible/fungible_vault.rs:267]: Let's rename to `is_feature_enabled_in_outer_object`.

File: radix-engine/src/system/system.rs [hunk 5]
```diff
@@ -640,39 +616,77 @@ where
...
-            blueprint,
-            &features,
-            &instance_schema,
-            fields,
-            kv_entries,
+        let blueprint_interface = self.get_blueprint_default_interface(
+            blueprint_id.package_address,
+            blueprint_id.blueprint_name.as_str(),
         )?;
+        let expected_blueprint_parent = blueprint_interface.outer_blueprint.clone();
```
Comment (iamyulong) [inline radix-engine/src/system/system.rs:630]: Rename `expected_blueprint_parent` to `expected_outer_blueprint`?

File: radix-engine/src/system/system.rs [hunk 6]
```diff
@@ -640,39 +616,77 @@ where
...
+        let (blueprint_info, associated_features) = if let Some(parent) = &expected_blueprint_parent
+        {
             match instance_context {
-                Some(context) if context.outer_blueprint.eq(parent) => Some(context.outer_object),
+                Some(context) if context.outer_blueprint.eq(parent) => {
+                    if !features.is_empty() {
+                        return Err(RuntimeError::SystemError(
+                            SystemError::InvalidChildObjectCreation,
+                        ));
+                    }
```
Comment (iamyulong) [inline radix-engine/src/system/system.rs:640]: Rename error to `InnerBlueprintFeatureSetNotSupported`

File: radix-engine/src/system/system.rs [hunk 7]
```diff
@@ -640,39 +616,77 @@ where
...
+
+                    let outer_object_info =
+                        self.get_object_info(context.outer_object.as_node_id())?;
+
+                    (
+                        ObjectBlueprintInfo::Inner {
+                            outer_object: context.outer_object,
+                        },
+                        outer_object_info.get_features(),
```
Comment (iamyulong) [inline radix-engine/src/system/system.rs:649]: Storing outer object features to inner object this way doesn't look right.

File: radix-engine-interface/src/blueprints/package/invocations.rs [hunk 8]
```diff
@@ -119,31 +129,13 @@ pub struct AuthConfig {
...
-        }
-    }
-
-    pub fn outer_auth(self) -> BTreeMap<MethodKey, MethodPermission> {
-        match self {
-            MethodAuthTemplate::Static { outer_auth, .. } => outer_auth,
-        }
-    }
+    Static(BTreeMap<MethodKey, MethodPermission>),
+    StaticUseOuterAuth(BTreeMap<MethodKey, MethodPermission>),
```
[the remark this answers is not in the document]
Comment (talekhinezh) [reply radix-engine-interface/src/blueprints/package/invocations.rs:133]: This is checked

**Pull request merged**